onnxruntime/core/framework/fallback_cpu_capability.cc

```diff
@@ -39,6 +43,132 @@ static bool IsSmallInitializer(const onnxruntime::GraphViewer& graph, const Node
+    //  2. finds `shape` is a shape-related variable since Reshape's 2nd input is a shape-related input,
+    //  3. and then records the producer of `shape` (i.e., `Concat`).
+    for (auto& input_index : shape_input_indices) {
+      auto input = node.InputDefs().at(input_index);
+      auto producer_node = graph.GetProducerNode(input->Name());
+      if (producer_node != nullptr && producer_node->OpType() != "Shape" && producer_node->OpType() != "Size") {
+        // Assume shape-computing sub-graphs begins with Shape, Size, or graph inputs.
+        // We should not fallback those nodes's upstream nodes to CPU; otherwise,
+        // it may change
+        //   GPU-tensor-x -> Mul -> GPU-tensor-y -> Shape -> CPU-tensor
+        // to
+        //   CPU-tensor-x -> Mul -> CPU-tensor -> Shape -> CPU-tensor
+        // and slows down the computation.
+
+        // After this for-loop, we will reversely traverse all nodes from every shape-producing node
+        // found here until hitting Shape, Size nodes, graph inputs, or graph initializers.
+        // All nodes on the traversal path will be forced to run on CPU.
+        LOGS_DEFAULT(INFO) << "Find a shape producing node (i.e., a node produces a tensor consumed as shape-like input in other nodes): " << node.Name() << std::endl;
+        shape_producing_nodes.insert(producer_node);
+      }
+    }
+  }
+
+  std::unordered_set<NodeIndex> shape_related_node_indices;
+  for (auto& node : shape_producing_nodes) {
+    LOGS_DEFAULT(INFO) << "Begin the (topologically reverse) traversing from shape producing node: " << node->Name() << std::endl;
+    std::vector<const Node*> start_nodes = {node};
+
+    auto to_stop = [](const Node* n1, const Node* n2) {
+      LOGS_DEFAULT(INFO) << "Skip the traversal from " << n1->Name() << " to " << n2->Name() << " since " << n2->Name() << " is a Shape or Size node." << std::endl;
+      return n2->OpType() == "Shape" || n2->OpType() == "Size";
+    };
+
+    // Reversely traverse all nodes from the shape-producing node.
+    // Force nodes to be run on CPU when all inputs and outputs are small.
+    // Stop the traversal when a "Shape" node is found.
+    graph.ReverseDFSFrom(
+        start_nodes,
+        [&shape_related_node_indices](const Node* n) {
+          LOGS_DEFAULT(INFO) << "Find an upstream node in shape sub-graph (let's fallback it to CPU): " << n->Name() << std::endl;
+          shape_related_node_indices.insert(n->Index());
+        },
+        nullptr,
+        NodeCompare(),
+        to_stop);
+  }
+
+  return shape_related_node_indices;
+}
+
+bool IsAggressiveCpuFallbackEnabled() {
+#if !defined(_WIN32) && ENABLE_TRAINING
+  // std::getenv is not available on each platform.
+  // Since ORT_AGGRESSIVE_CPU_FALLBACK is experimental,
+  // we only allow it for training to avoid build issues on
+  // custom platform such as XBox.
+  const char* p_env_var = std::getenv("ORT_AGGRESSIVE_CPU_FALLBACK");
```

Comment:
please don't use env variables. use session options instead. https://github.com/microsoft/onnxruntime/blob/77da2ef278a4e77cca4cef4e5d72ed1ef46fcce3/include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h

Reply:
It's inside execution provider, which don't have access to session options.

Reply:
The issue being discussed: this function does not have access to session_options, but it is being called inside every provider, so it is a lot of refactoring.

On the other hand,  env var is a logistical (many places in diff parts of the code) and security problem.

Reply:
Let's make the necessary changes to avoid the usage of env variables.

Reply:
I looked into it and I suggest we add a member to the ExecutionProvider base class that will hold a ptr to session_options member of the owening session.  Then each EP will have an access.
The best place to set it is at `InferenceSession::RegisterExecutionProvider` where various provider properties are set.

Reply:
Can ProviderOption be used instead of SessionOption ? - This seems like a provider level abstraction rather than a session level abstraction.

Reply:
Given that it's the EP that is making the decision to call ```GetCpuPreferredNodes()``, it does make sense to make this a provider option. 
EP configs are in a bit of a mixed state currently. Some EPs are consuming them from session options and some from provider options. For e.g. see [this](https://github.com/microsoft/onnxruntime/blob/0c5f4924937a6d9556e99a3fb59b33f9cc7b35a6/onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h#L20-L25).

Reply:
we've been trying to go the other direction. options that can be shared by all ep's we've been pushing them to session options. e.g. https://github.com/microsoft/onnxruntime/pull/18877
it doesn't make sense for ep's to duplicate the same option.

Reply:
Let's continue the discussion offline.

Reply:
It's in SessionOptions now.